query-engine/nodejs-drivers/src/driver.rs

```diff
@@ -47,20 +50,97 @@ pub fn reify(js_driver: JsObject) -> napi::Result<Driver> {
     Ok(driver)
 }
 
+// This result set is more convenient to be manipulated from both Rust and NodeJS.
+// Quaint's version of  ResultSet is:
+//
+// pub struct ResultSet {
+//     pub(crate) columns: Arc<Vec<String>>,
+//     pub(crate) rows: Vec<Vec<Value<'static>>>,
+//     pub(crate) last_insert_id: Option<u64>,
+// }
+//
+// If we used this ResultSet would we would have worse ergonomics as quaint::Value is a structured
+// enum and cannot be used directly with the #[napi(Object)] macro. Thus requiring us to implement
+// the FromNapiValue and ToNapiValue traits for quaint::Value, and use a different custom type
+// representing the Value in javascript.
+//
 #[napi(object)]
-#[derive(Debug, Serialize, Deserialize)]
+#[derive(Debug)]
 pub struct ResultSet {
-    pub columns: Vec<String>,
+    pub column_types: Vec<ColumnType>,
+    pub column_names: Vec<String>,
+    // Note this might be encoded differently for performance reasons
+    pub rows: Vec<Vec<serde_json::Value>>,
+}
+
+#[napi]
+#[derive(Debug)]
+pub enum ColumnType {
+    Int32,
+    Int64,
+    Float,
+    Double,
+    Text,
+    Enum,
+    Bytes,
+    Boolean,
+    Char,
+    Array,
+    Numeric,
+    Json,
+    DateTime,
+    Date,
+    Time,
+}
 
-    // TODO: support any JS-serializable type, not just String.
-    pub rows: Vec<Vec<String>>,
+#[napi(object)]
+#[derive(Debug)]
+pub struct QueryParameters {
+    pub sql: String,
+    pub parameters: Vec<serde_json::Value>,
+}
+
+impl From<ResultSet> for QuaintResultSet {
+    fn from(mut val: ResultSet) -> Self {
+        // TODO: extract, todo: error rather than panic?
+        let to_quaint_row = move |row: &mut Vec<serde_json::Value>| -> Vec<quaint::Value<'static>> {
+            let mut res = Vec::with_capacity(row.len());
+
+            for i in 0..row.len() {
+                match val.column_types[i] {
+                    ColumnType::Int64 => match row.remove(0) {
```

Comment:
This will likely need to be extracted and optimized. For each row, each value will need to go through a conversion function, but the final conversion function will be the same for the same column of every row.